[SPIRV] Audit `select` Result in SPIRVInstructionSelector

inbelic opened the pull request:
- as per the definition of `select` in GlobalISel/InstructionSelector.h the return value is a boolean denoting if the select was successful
- doing `Result |=` is incorrect as all inserted instructions should be succesful, hence we change to using `Result &=`
- ensure that the return value of all BuildMI instructions are propagated correctly

inbelic committed:
[SPIRV] Audit `select` Result in SPIRVInstructionSelector

- as per the definition of `select` in GlobalISel/InstructionSelector.h
the return value is a boolean denoting if the select was successful

- doing `Result |=` is incorrect as all inserted instructions should be
succesful, hence we change to using `Result &=`

inbelic committed:
Merge branch 'main' into inbelic/spirv-audit-result

pow2clk reviewed (commented):
These changes all look good. There are a few instances of BuildMI being called without the result being read or otherwise used that you might consider while you're at it. I don't know why they would fail, but I don't know specifically why any of these would fail, but if they do, it should be passed up.

inbelic committed:
additionally check all uses of BuildMI

- propogate all BuildMI returns up

propogate the buildmi result in buildI32Constant

inbelic committed:
Merge branch 'main' into inbelic/spirv-audit-result

inbelic commented:
Merged to get the splat test failure fix

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1038,12 +1040,13 @@ bool SPIRVInstructionSelector::selectMemOperation(Register ResVReg,
-        .addImm(SPIRV::StorageClass::UniformConstant)
-        .addUse(Const)
-        .constrainAllUses(TII, TRI, RBI);
+    Result &=
```
I would just declare Result here.

spall replied:
Oh this is inside an if statement ignore me completely.

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1068,30 +1073,34 @@ bool SPIRVInstructionSelector::selectAtomicRMW(Register ResVReg,
-  Register ScopeReg = buildI32Constant(Scope, I);
+  auto ScopeConstant = buildI32Constant(Scope, I);
+  Register ScopeReg = ScopeConstant.first;
+  Result &= ScopeConstant.second;
```
I would again declare Result here

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1205,21 +1219,22 @@ bool SPIRVInstructionSelector::selectOverflowArith(Register ResVReg,
   }
   // Build boolean value from the higher part.
-  Status &= BuildMI(BB, I, I.getDebugLoc(), TII.get(SPIRV::OpINotEqual))
+  Result &= BuildMI(BB, I, I.getDebugLoc(), TII.get(SPIRV::OpINotEqual))
```
I would do 'return Result && BuildMI', instead of returning later.

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1229,17 +1244,26 @@ bool SPIRVInstructionSelector::selectAtomicCmpXchg(Register ResVReg,
-    ScopeReg = buildI32Constant(Scope, I);
+    auto ScopeConstant = buildI32Constant(Scope, I);
+    ScopeReg = ScopeConstant.first;
+    Result &= ScopeConstant.second;
```
ditto

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1618,26 +1642,27 @@ bool SPIRVInstructionSelector::selectAnyOrAll(Register ResVReg,
-        .addUse(InputRegister)
-        .addUse(ConstZeroReg)
-        .constrainAllUses(TII, TRI, RBI);
+    Result &= BuildMI(BB, I, I.getDebugLoc(), TII.get(SpirvNotEqualId))
```
ditto

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1796,7 +1821,7 @@ bool SPIRVInstructionSelector::selectDot4AddPackedExpansion(
     // A[i]
     Register AElt = MRI->createVirtualRegister(&SPIRV::IDRegClass);
-    Result |= BuildMI(BB, I, I.getDebugLoc(), TII.get(ExtractOp))
+    Result &= BuildMI(BB, I, I.getDebugLoc(), TII.get(ExtractOp))
```
ditto

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -2194,9 +2220,9 @@ SPIRVInstructionSelector::buildI32Constant(uint32_t Val, MachineInstr &I,
                .addImm(APInt(32, Val).getZExtValue());
     }
-    constrainSelectedInstRegOperands(*MI, TII, TRI, RBI);
+    Result &= constrainSelectedInstRegOperands(*MI, TII, TRI, RBI);
```
ditto

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -2712,16 +2744,22 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
         buildI32Constant(SPIRV::MemorySemantics::SequentiallyConsistent, I);
-    Register ScopeReg = buildI32Constant(SPIRV::Scope::Workgroup, I);
+    Register MemSemReg = MemSemConstant.first;
+    Result &= MemSemConstant.second;
```
ditto

spall reviewed (commented):
Wow its hard to see the nesting on some of these with the diffs.

inbelic committed:
review comment:

- make result return consistent to use return Result && when applicable

pow2clk commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1229,17 +1244,26 @@ bool SPIRVInstructionSelector::selectAtomicCmpXchg(Register ResVReg,
+    else {
+      auto MemSemNeqConstant = buildI32Constant(MemSemEq, I);
+      MemSemNeqReg = MemSemNeqConstant.first;
+      Result &= MemSemNeqConstant.second;
```
I'm not going to insist on it, but just FYI, you could have made `buildI32Constant` return `Register()` in the failing case, which could be identified by checking `Register.isValid()` rather than returning an additional bool. It might have made this a little more elegant, but I'm not sure about that and it's entirely a non-functional change, so I'll happily defer to the implementer here.

inbelic replied:
I agree it would be more elegant.

But from scanning the file, there doesn't seem to be a consistent precedent to checking a register's validity before use. In particular the returns of similar functions `buildZerosVal` and `BuildOnesVal`. I think to differentiate from those functions, and to ensure the result is checked in future calls we should explicitly return the type.

This might be a good reason to also audit the use of returned Registers, which I think is out of scope for this pr. Then we could switch to returning an invalid Register.

So I will merge as-is.

pow2clk reviewed (approved):
Looks good!

I was reminded that have a half-finished change that made a similar cleanup attempt, but it wasn't as complete as what you did.
